File: engine/grid/map.py
"""Grid map helpers."""

from __future__ import annotations

from engine.core.state import BattleState


def in_bounds(state: BattleState, x: int, y: int) -> bool:
    return 0 <= x < state.battle_map.width and 0 <= y < state.battle_map.height


def is_blocked(state: BattleState, x: int, y: int) -> bool:
    return (x, y) in set(state.battle_map.blocked)


def is_occupied(state: BattleState, x: int, y: int) -> bool:
    for unit in state.units.values():
        if unit.alive and unit.x == x and unit.y == y:
            return True
    return False

File: src/grid/map.ts
/**
 * Grid map helpers.
 */

import { BattleState, unitAlive } from "../engine/state";

export function inBounds(state: BattleState, x: number, y: number): boolean {
  return x >= 0 && x < state.battleMap.width && y >= 0 && y < state.battleMap.height;
}

export function isBlocked(state: BattleState, x: number, y: number): boolean {
  return state.battleMap.blocked.some(([bx, by]) => bx === x && by === y);
}

export function isOccupied(state: BattleState, x: number, y: number): boolean {
  return Object.values(state.units).some(
    (unit) => unitAlive(unit) && unit.x === x && unit.y === y,
  );
}

/** PF2e 5ft grid: convert a distance in feet to a tile radius. */
export function tilesFromFeet(feet: number): number {
  return Math.max(1, Math.floor((feet + 4) / 5));
}
